### PR TITLE
[CR]NPC AI update + cleanup

### PR DESCRIPTION
**Pull request by Coolthulhu**
New features or re-enabled old features:
- NPCs can use healing items. Not just hardcoded bandage and first aid, but any of them that use "heal" item actor
- NPCs should be a tiny bit more sane about weapon selection
- Re-enabled NPCs picking up items on their own. For some reason they stopped doing so
- Friendly NPCs don't pick up items by default
- Improved NPC speed estimates, especially against player
- After picking up items, NPCs will check if those items could be good weapons

Code quality changes:
- Removed special cases for player seeing NPC but not their adjacent target or only the target and not the NPC. Those were big `if` trees with very minor effects
- Added `Creature::get_name`. It's like `disp_name`, except doesn't add "the" in case of monsters.
- Added an `ai_cache` structure which currently contains `target`, `danger`, `total_danger` and `danger_assessment`. Those values were passed around and keeping track of them was annoying at times. Later it may contain some stats about environment, for example nearby dangerous monsters (not just the strongest one).
- New `wield_best_weapon` uses a different structure from most NPC code: it wields best weapon immediately. Most old NPC code instead decides to do something, then "reconstructs" that something later on - for example, decides to wield a weapon, then later figures out which. This can be hard to use without a good way to remember what was decided first, so I'd move away from it unless it's necessary to keep it
- Added a `npc_noop` fake action to help with the above. NPC will do a no-op after some change of state that requires reconsidering some important stuff. For example, NPC running out of patience and going hostile or NPC wielding a weapon. This is just an AI no-op to inform the `execute_action` that something was done, wielding will still drain moves

### CODE REVIEW

**Comment by mugling** [comment]
It's good to see the NPC code getting lots of attention as of recent.

**Comment by Malkeus** [comment]
Do these changes apply to static npc's and random npc's equally?

**Comment by kevingranade** [comment]
@Malkeus yes, all the AI code is the same, it's just that random NPCs are much more likely to want to wander around and/or be hostile.

File: src/npcmove.cpp [hunk 1]
```diff
@@ -2632,13 +2582,23 @@ bool npc::complain()
...
         say( "<yawn>" );
         complaints[fatigue_string] = calendar::turn;
         return true;
     }
 
+    // Radiation every 10 minutes
+    if( radiation > 90 &&
+        complaints[radiation_string] < calendar::turn - MINUTES(10) &&
+        (rules.allow_complain || radiation > 150) ) {
+        say( _("I'm suffering from radiation sickness...") );
```
Comment (kevingranade) [inline src/npcmove.cpp]: There's a lot of room for this to be a gruesome list of symptoms rather than a simple observation.

**Pull request merged**